### ISSUE TITLE
Consider relicensing to a permissive license

### ISSUE THREAD

**Issue by TheButlah** [p-avital/token-cell-rs#3]
Hi! The EPL 2.0 is copyleft/viral. We mix our (MIT/Apache) source code with proprietary third party vendor SDKs. These vendor sdks are code that we don't have legal rights to open source, and which are incompatible with EGPL-2.0. This is quite common in firmware / code that deals with hardware unfortunately :(

Would you consider re-licensing to a permissive license, such as an MIT/Apache 2.0 license so that more projects can make use of this crate?

**Comment by p-avital** [comment]
Hi Ryan,

If you don't mind, I'll treat this thread as the global one for all other repos to avoid spam.

To my knowledge, EPL-2.0 should be permissive enough for your usage: it fully permits usage of three software in commercial products, only requiring that _changes to the software itself_ be contributed back.

Unlike GPL, its definition of derivative works is much more narrow.

You are free to use and distribute the software however you see fit (possibly vendoring it). Only if your usage requires modification (to support additional platforms for example), would you then be required to open source your fork of the repo (and nothing more).

This notably means that if your usage of these libraries is simply "they're in our dependency tree", I expect you needn't do anything; and nor does the licensing need to change, as EPL-2.0 and MIT/Apache are essentially equivalent for that use case.

Are these terms really not agreeable to you?

**Comment by TheButlah** [comment]
You are correct I think, the EPL is similar to the MPL in the sense that distributing it unmodified does not trigger the copyleft permissions, that definitely helps. For example, my organization allows MPL licensed code, on a case-by-case basis, although its frowned upon.

However, I think there are a few concerning provisions in the EPL that go a lot further than a license like the MPL:
* Section 4 broadens the notion of disclaimer of warranty. Most licenses disclaim warranty on behalf of the authors. But the EPL goes a lot further, and compels users of the software to further pay and defend such authors, *if* those authors can claim that the lawsuit against the author was because of the corporate user's advertisement. I think this is an unusual clause in a software license and could easily be abused. Also, it subjects average people to the possiblity of paying legal fees on behalf of third parties.
* The license is not clear what "Commercial" means, as far as I can tell it isn't defined at all. For example, is someone accepting donations for their work that uses an EPL dependency providing a commercial offering? Would they then be able to accidentally get themselves into a situation where they have to pay legal fees on behalf of a legal battle they ought not to be involved in to begin with?
* The license takes an overly broad definition of "Contributor" . When we think of a "contributor" we typically mean someone that has modified covered source code. But the EPL defines "contributor" as anyone that *distributes* the program or source code. I think that this definition is overly broad and leads to confusion. Ultimately as *definitely not a lawyer*, I am not confident I understand what the EPL means sufficiently - I will need to consult a lawyer.

Because of these issues, I think most organizations would still need to consult with a legal team to explicitly approve this license, and many legal teams would not be willing to add the liability of defending third parties - which makes adoption harder.

I wonder, if the focus is specifically on ensuring that modifications but not use of the code triggers copyleft, maybe you could consider using MPL? I think that MPL, while not as permissive as MIT or Apache, would satisfy your desire to ensure contributions are shared back to the project, without having the additional problematic clauses from above.

**Comment by TheButlah** [comment]
See also: https://github.com/eclipse-zenoh/zenoh/issues/1625#issuecomment-2528948963

TLDR:
* I was wrong about copyleft, EPL-2.0 is similar to MPL in that simply consuming code as a dependency and not modifying it doesn't trigger copyleft. This alleviates most of the concerns.
* The indemnity clause is still a bit problematic, but the risk is fairly low and we don't think we would ever run afoul of this, so for my particular company we were fine with it.

I will leave this issue and the others open, but feel free to close them if its something that is not planned.
